### DISCUSSION TITLE
Consider to native support for WASI and Webassembly？

### DISCUSSION THREAD

**Discussion by JermineHu** [vlang/v#11586] [Ideas]
Reference to the following two items: 

https://github.com/fastly/lucet

https://github.com/wasmerio/wasmer

**Comment by andreabravetti** [comment]
It wuold be usefull to have full GUI support with wasm inside a browser

**Comment by syrusakbary** [comment]
I would love to see V supporting WebAssembly/WASI as a target, with first class support in @wasmerio.
Let us know if we can help in any way! :)

**Comment by medvednikov** [comment]
Yes, that would be great.

**Comment by tomByrer** [comment]
Any progress on WASM output please?

**Comment by sanathusk** [comment]
I would love to see V lang support web assembly.

**Comment by richardanaya** [comment]
Hey all, I work on a library specifically for enabling WebAssembly languages to communicate with javascript. 

http://wasm.js.org/

It might be a way to help accelerate efforts in V to expose more of the host.

**Comment by Jengro777** [comment]
Taking wasi as the compilation target will bring us a lot

**Comment by JalonSolov** [comment]
A new WASM backend was merged into V last week in PR #17368

You can now compile to WASM and run the code with wasmer
```
$ v -b wasm examples/wasm/hello_world.v
$ wasmer examples/wasm/hello_world.wasm
Hello WASI!
$
```
or compile to WASM and run in the browser (see instructions in `examples/wasm/mandelbrot` directory).

It isn't 100% complete, but there have already been a couple more PRs merged since then to improve the WASM support.

**Reply by syrusakbary** [reply]
That's awesome @JalonSolov!

**Reply by Jengro777** [reply]
why wasmer ？
How about wasmtime and wasmedge？

**Reply by JalonSolov** [reply]
... because that's what the person was familiar with that implemented the backend?  You are free to try those other options, and if they work, the instructions can be updated to list them as options.

**Reply by Jengro777** [reply]
![image](https://user-images.githubusercontent.com/35259040/226226598-e51ab10a-be46-41f2-999a-8c857641d76c.png)

Why means？

**Reply by JalonSolov** [reply]
Looks like binaryen didn't get installed correctly.  Perhaps it just needs the execute bit set.

**Comment by Jengro777** [comment]
![image](https://user-images.githubusercontent.com/35259040/226781408-f20d3854-1561-4d4c-a0e0-74483cf6beeb.png)

This is trouble.

**Reply by JalonSolov** [reply]
In English?  Since you posted a screenshot, I can't run it through a translator.

**Reply by getfile** [reply]
it say "This app can't run on your PC"

**Reply by JalonSolov** [reply]
Ok, but that doesn't tell me anything other than that you can't run something on that machine.  Does it mention binaryen specifically?  Are you certain you have a version of binaryen that is _supposed_ to run on your machine?

**Reply by Jengro777** [reply]
![image](https://user-images.githubusercontent.com/35259040/227071563-c1d638cb-4053-4fa9-9393-0bf4ab1e5dea.png)

binaryen version_ Is 112 incompatible with the Win10 system? "I operate everything according to the official v documentation, but after installing binaryen, executing 'v - b wasm hello_world. v' will prompt 'This app cannot run on your computer', and 'To find a version that works for your computer, please consult the software publisher'.".



This makes me very confused. Is it because binaryen is not compatible with the Win10 system?

**Reply by JalonSolov** [reply]
Can you change your directory to `thirdparty/binaryen/bin` and check to see if the execute bit is set on all the files?  Also, check the file type, and make sure you got the Windows executables rather than something else.

**Reply by Jengro777** [reply]
![image](https://user-images.githubusercontent.com/35259040/227128354-33405093-c449-44e8-b767-8d664537aa03.png)

**Reply by JalonSolov** [reply]
I was able to try this in a Windows 10 vm, and got similar problems.  Apparently this was not tested on Windows, or the tester should have seen the same.

I verified that the binaryen exes that are retrieved do run as expected, V just doesn't run them correctly, somehow.

@l1mey112

**Reply by Wajinn** [reply]
Hope l1mey112 is OK.  Happen to check on their site, and it's down.

**Reply by Jengro777** [reply]
https://github.com/l1mey112/l1mey112  is alrady empty.

https://github.com/l1mey112/binaryen-v  is normal.

Will someone fix this bug now?

**Reply by Wajinn** [reply]
Edit: Good to see that l1mey112 is back and their website is back up.

**Reply by Jengro777** [reply]
I also tested it on debian11, and the following error will be prompted. I don't think this should be the case by default

```
root@avey:/opt/v# v -b wasm examples/wasm/hello_world.v 
cannot compile `/opt/v/cmd/tools/builders/wasm_builder.v`: 
==================
tcc: error: library 'stdc++' not found
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

**Reply by l1mey112** [reply]
@Avey777 Hello!

The experimental WebAssembly backend currently has issues with binaryen, it's packaging. The WebAssembly backend has not been tested with windows, usually library support is patchy there. As for Debian, it is known that Debian does not contain an up to date `libstd++`. Even if you had installed it, it would have not worked. To install `libstdc++`, try installing a C++ compiler such as `g++`.

I will try to get back to you if you have further questions.

**Reply by km19809** [reply]
I have a similar problem. My v cannot find binaryen library, after installing it. I am using Windows 10 and v 0.3.3.
```
cannot compile `C:\v\cmd\tools\builders\wasm_builder.v`:
==================
C:/Users/km198/AppData/Local/Temp/v_0/wasm_builder.13306416624556646702.tmp.c:67727: warning: assignment from incompatible pointer type
tcc: error: library 'binaryen' not found
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

**Reply by JalonSolov** [reply]
Try running `v cmd/tools/install_binaryen.vsh` to get that installed in `thirdparty/binaryen`.

May still not work for you, but at least it should give the same error others are seeing.

**Reply by km19809** [reply]
@JalonSolov 
It is weird because I already ran it. 
Also, I can run `binaryen_unittest.exe`, which succeeds in all the tests. 
![image](https://user-images.githubusercontent.com/20869932/227788726-9d7b4ea5-7640-4df2-bba0-8002fdaf323c.png)

**Reply by JalonSolov** [reply]
Definitely weird, but as stated already... this is a new backend, and hasn't been tested much on Windows - at all - so there are bound to be issues.

**Reply by spytheman** [reply]
Try with `-cc msvc`. The CI has a working job with it, see:
<https://github.com/vlang/v/blob/master/.github/workflows/wasm_backend_tests_ci.yml#L97>
